servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -416,12 +416,19 @@ private Single<C> selectConnection0(final Predicate<C> selector, @Nullable final
                     // Invoke the selector before adding the connection to the pool, otherwise, connection can be
                     // used concurrently and hence a new connection can be rejected by the selector.
                     if (!selector.test(newCnx)) {
-                        // Failure in selection could be temporary, hence add it to the queue and be consistent
-                        // with the fact that select failure does not close a connection.
-                        return newCnx.closeAsync().concat(failed(StacklessConnectionRejectedException.newInstance(
+                        // Failure in selection could be the result of connection factory returning cached connection,
+                        // and not having visibility into max-concurrent-requests, or other threads already selected the
+                        // connection which uses all the max concurrent request count.
+
+                        // Just in case the connection is not closed add it to the host so we don't lose track,
+                        // duplicates will be filtered out.
+                        host.addConnection(newCnx);
```

Comment:
It looks like we can still leak the connection if `addConnection` method returns `false`